lib/puma/cluster/worker.rb

```diff
@@ -84,25 +89,34 @@ def run
               idx = idx.to_i
               if idx == -1 # stop server
                 if restart_server.length > 0
-                  restart_server.clear
-                  server.begin_restart(true)
-                  @config.run_hooks(:before_refork, nil, @log_writer, @hook_data)
+                  # acquire a mutex to ensure the server is started before stopping
+                  mutex.synchronize do
+                    restart_server.clear
+                    server.begin_restart(true)
+                    @config.run_hooks(:before_refork, nil, @log_writer, @hook_data)
+                  end
                 end
               elsif idx == 0 # restart server
-                restart_server << true << false
-              else # fork worker
-                worker_pids << pid = spawn_worker(idx)
-                @worker_write << "#{Puma::Const::PipeRequest::FORK}#{pid}:#{idx}\n" rescue nil
+                restart_server << Puma::Const::WorkerCmd::RESTART << Puma::Const::WorkerCmd::STOPPED
+              else
+                # spawn new worker
+                restart_server << "#{Puma::Const::WorkerCmd::SPAWN}#{idx}"
```

Comment:
We should definitely pass the idx state somewhere else, this is allocating multiple strings every loop for objects that are otherwise interned/intern-able.